Expose allowsFollowingLink/allowsFollowingImageURL  to _WKContextMenuElementInfo

arenevier opened the pull request:
#### 810cd6f9cfa757676ddc5d30ffac922b1532709c
<pre>
Expose allowsFollowingLink/allowsFollowingImageURL  to _WKContextMenuElementInfo
<a href="https://bugs.webkit.org/show_bug.cgi?id=293025">https://bugs.webkit.org/show_bug.cgi?id=293025</a>
<a href="https://rdar.apple.com/150874653">rdar://150874653</a>

Reviewed by Alex Christensen.

This will allow the UI process to get more informations when a context menu is shown

allowsFollowingLink will be true iff the context menu is over a link
that can be followed. allowsFollowingImageURL will be true iff it&apos;s over
an image whose URL can be followed

* Source/WebCore/page/ContextMenuContext.cpp:
(WebCore::ContextMenuContext::ContextMenuContext):
* Source/WebCore/page/ContextMenuContext.h:
(WebCore::ContextMenuContext::allowsFollowingLink const):
(WebCore::ContextMenuContext::allowsFollowingImageURL const):
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::allowsFollowingLink const):
(WebCore::HitTestResult::allowsFollowingImageURL const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/ContextMenuContextData.cpp:
(WebKit::ContextMenuContextData::ContextMenuContextData):
* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::allowsFollowingLink const):
(WebKit::ContextMenuContextData::allowsFollowingImageURL const):
* Source/WebKit/Shared/ContextMenuContextData.serialization.in:
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::allowsFollowingLink const):
(API::HitTestResult::allowsFollowingImageURL const):
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo allowsFollowingLink]):
(-[_WKContextMenuElementInfo allowsFollowingImageURL]):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingImageURLTrue)):
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingImageURLFalse)):
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingLinkTrue)):
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingLinkFalse)):

Canonical link: <a href="https://commits.webkit.org/294987@main">https://commits.webkit.org/294987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3706d218cd36862618ec310b0257ce92270c2044

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13753 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/108921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/54383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105771 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31979 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/108921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/54383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106737 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/18449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/93575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/108921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11623 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53759 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/11681 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111311 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30887 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89776 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/87465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/32342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16834 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30815 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/36118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/30609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/33944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -219,6 +220,32 @@ bool HitTestResult::isSelected() const
+    if (linkURL.isEmpty())
+        return false;
+
+    auto* innerFrame = innerNodeFrame();
```
auto* should be RefPtr

arenevier replied:
That code is similar to HitTestResult::hasEntireImage. Should we modify ::hasEntireImage to also use a RefPtr?

achristensen07 commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -219,6 +220,32 @@ bool HitTestResult::isSelected() const
+    if (!innerFrame)
+        return false;
+
+    return innerFrame->protectedDocument()->protectedSecurityOrigin()->canDisplay(linkURL, OriginAccessPatternsForWebProcess::singleton());
```
I think we should null check protectedDocument()

achristensen07 commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -219,6 +220,32 @@ bool HitTestResult::isSelected() const
+    if (linkURL.isEmpty())
+        return false;
+
+    auto* innerFrame = innerNodeFrame();
```
Same in this function

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
```diff
@@ -38,6 +38,9 @@ WK_CLASS_AVAILABLE(macos(10.12))
 @property (nonatomic, readonly, copy, nullable) NSString *qrCodePayloadString WK_API_AVAILABLE(macos(14.0));
 @property (nonatomic, readonly) BOOL hasEntireImage WK_API_AVAILABLE(macos(14.0));
 
```
nit: I would remove this newline.

cdumez commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -219,6 +220,40 @@ bool HitTestResult::isSelected() const
+    if (!innerFrame)
+        return false;
+
+    RefPtr document = innerFrame->protectedDocument();
```
`RefPtr document = innerFrame->document();`

We do not call `protectedX()` variants to initialize variables. The name is longer and there is no benefit.

cdumez commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -219,6 +220,40 @@ bool HitTestResult::isSelected() const
+    if (!innerFrame)
+        return false;
+
+    RefPtr document = innerFrame->protectedDocument();
```
Ditto.

achristensen07 commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -57,6 +57,7 @@
 #include "VisibleUnits.h"
 #include "XLinkNames.h"
 #include <wtf/TZoneMallocInlines.h>
+#include "OriginAccessPatterns.h"
```
This is out of order.  It should be above Page.h.

Commit:
Expose allowsFollowingLink/allowsFollowingImageURL  to _WKContextMenuElementInfo
https://bugs.webkit.org/show_bug.cgi?id=293025
rdar://150874653

Reviewed by Alex Christensen.

This will allow the UI process to get more informations when a context menu is shown

allowsFollowingLink will be true iff the context menu is over a link
that can be followed. allowsFollowingImageURL will be true iff it's over
an image whose URL can be followed

* Source/WebCore/page/ContextMenuContext.cpp:
(WebCore::ContextMenuContext::ContextMenuContext):
* Source/WebCore/page/ContextMenuContext.h:
(WebCore::ContextMenuContext::allowsFollowingLink const):
(WebCore::ContextMenuContext::allowsFollowingImageURL const):
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::allowsFollowingLink const):
(WebCore::HitTestResult::allowsFollowingImageURL const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/ContextMenuContextData.cpp:
(WebKit::ContextMenuContextData::ContextMenuContextData):
* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::allowsFollowingLink const):
(WebKit::ContextMenuContextData::allowsFollowingImageURL const):
* Source/WebKit/Shared/ContextMenuContextData.serialization.in:
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::allowsFollowingLink const):
(API::HitTestResult::allowsFollowingImageURL const):
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo allowsFollowingLink]):
(-[_WKContextMenuElementInfo allowsFollowingImageURL]):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingImageURLTrue)):
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingImageURLFalse)):
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingLinkTrue)):
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuElementInfoAllowsFollowingLinkFalse)):

Canonical link: https://commits.webkit.org/294987@main
